### PR TITLE
guix: Make source tarball using git-archive

### PR DESCRIPTION
**Pull request by dongcarl**
Based on: #18556
Related: https://github.com/bitcoin/bitcoin/pull/17595#discussion_r399728721

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

**Comment by MarcoFalke** [comment]
You'll also need to adjust this line in guix for the correct naming:

```
            cp --target-directory=unsigned/ "$OUTDIR"/bitcoin-*-setup-unsigned.exe

**Review by maflcko — APPROVED** [review]

**Comment by MarcoFalke** [comment]
Needs rebase

File: contrib/gitian-descriptors/assign_DISTNAME [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#
+# A helper script to be sourced into the gitian descriptors
+
+if RECENT_TAG="$(git describe --exact-match HEAD)"; then
+    VERSION="${RECENT_TAG#v}"
+else
+    VERSION="$(git rev-parse --short HEAD)"
```
Comment (MarcoFalke) [inline contrib/gitian-descriptors/assign_DISTNAME:10]:
```suggestion
    VERSION="$(git rev-parse --short=10 HEAD)"
```

Not sure what the git logic is, but I couldn't verify from the doc that this is deterministic when the git version changes (and/or the number of objects in the repo)

**Comment by MarcoFalke** [comment]
Also in the guix readme. Could you remove the requirement that guix needs two cores. I am running this on half a core and everything is fine.

**Comment by MarcoFalke** [comment]
```

 /bin/mkdir -p '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32/installed/bitcoin-df4620517b//lib/pkgconfig'
 /usr/bin/install -c -m 644 libbitcoinconsensus.pc '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32/installed/bitcoin-df4620517b//lib/pkgconfig'
make[2]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32'
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-w64-mingw32'
+ cp -f './bitcoin-*-win64-setup-unsigned.exe' /home/ubuntu/out/bitcoin-df4620517b-win64-setup-unsigned.exe
cp: cannot stat './bitcoin-*-win64-setup-unsigned.exe': No such file or directory

**Comment by dongcarl** [comment]
Performed builds of `27e63e01c` locally:

Gitian Linux:
```
7767a6d95964525457dd096554b703ec0d962c994afc817672834fce5ad360c1  bitcoin-27e63e01c-aarch64-linux-gnu-debug.tar.gz
29c4782d005041a8aaf0dc0e4a23b9d9ada7fe14649be20c7306e25ebc8409b6  bitcoin-27e63e01c-aarch64-linux-gnu.tar.gz
60143f3c6e890dd946c4043a0a6b3d7f7dfaad995430bfcfb967f3c65e1ed1f3  bitcoin-27e63e01c-arm-linux-gnueabihf-debug.tar.gz
d1f12f2c46ec4de005f66c81c480edf671c1dee69b28d4768324723b8174cb9a  bitcoin-27e63e01c-arm-linux-gnueabihf.tar.gz
ae589966cd6f060b6dd2e82d75cd8db8b069ced215b0ffe71874729568da2838  bitcoin-27e63e01c-riscv64-linux-gnu-debug.tar.gz
f192b694117c5ced00e7d039ff653eca6cf26e034d7f413375dd09d3bfd7f95f  bitcoin-27e63e01c-riscv64-linux-gnu.tar.gz
4aa2d36209d47d930946d53103974eacf1d666ffe3bc27817a205e2b5a070acf  bitcoin-27e63e01c-x86_64-linux-gnu-debug.tar.gz
b0535037e28aadcf826dbe7a9dc49274aa3ab8f3885ce6b0365940a13c1b7c67  bitcoin-27e63e01c-x86_64-linux-gnu.tar.gz
a8daae18536b433d05ae7972355fc99d535f2687f2e078d1a4a0893cc2c6bf5f  src/bitcoin-27e63e01c.tar.gz
ad9206e894818fef5d3e2c5674ef01b1f5f26ac0a820f61d051b76a6759a3ee1  bitcoin-core-linux-0.21-res.yml
```

Gitian Win:
```
a4288e594f4bcd623b6fd43b041db2530868be4eef27a9cbeab978c1ba5a4972  bitcoin-27e63e01c-win-unsigned.tar.gz
8a9d2255b7a790b73554e15c189b306f431826f506350a2a5daabc367fe32269  bitcoin-27e63e01c-win64-debug.zip
9ce0ffcfb20213b547721b43beceff4ab4f8ffb9aed289d945a5cdbf0070e4bc  bitcoin-27e63e01c-win64-setup-unsigned.exe
aa0427ba6c1b3ceb2343d25f07daf6191af7c1c27efac44be754af628aae1688  bitcoin-27e63e01c-win64.zip
a8daae18536b433d05ae7972355fc99d535f2687f2e078d1a4a0893cc2c6bf5f  src/bitcoin-27e63e01c.tar.gz
93f7321989ba1d7268d8d1f0d37d0a135e281e834a070fe9eb936e1d4d62d542  bitcoin-core-win-0.21-res.yml
```

All available Guix architectures:
```
$ find output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
25aeed36ffc1eba8aff7e12f62011575e2270f07f8c25d14ffb3d0691996d919  output/bitcoin-27e63e01c-aarch64-linux-gnu-debug.tar.gz
7423aad2b0306a828b1dbe7eeb51c22c301298d9a6ae409c53546433374c8e29  output/bitcoin-27e63e01c-aarch64-linux-gnu.tar.gz
f8f027993c373698fb5f1b868769f903e246594eba69fdfe2861508828fecf0c  output/bitcoin-27e63e01c-arm-linux-gnueabihf-debug.tar.gz
6e4c079f55024348237cd55f394db1d09a1b91b8f520e64b025984d6bb7bddb0  output/bitcoin-27e63e01c-arm-linux-gnueabihf.tar.gz
7159dc6c2acabcf089b37a8c618e8558f68a44b24a733861c2b88c865c72b3f5  output/bitcoin-27e63e01c-riscv64-linux-gnu-debug.tar.gz
a85afdeff32b003f11a60831b6654a4fb1fb02b899f10392006f33fbaf2b3e4a  output/bitcoin-27e63e01c-riscv64-linux-gnu.tar.gz
273958496232e58fc4204b28947e8b10ebd0244f7016d0098b391574639aceca  output/bitcoin-27e63e01c-win-unsigned.tar.gz
49844ea00fcaebb141cc5b0e6a0a135df045dec1bb79256ef0f65e2a11d8ad1d  output/bitcoin-27e63e01c-win64-debug.zip
2c7c70a33492f17a8f35c6c7348a10ad795e85855a929196d499956b048d09a2  output/bitcoin-27e63e01c-win64-setup-unsigned.exe
8985a7d8d0b8511c2e36e050c0f006346ef66c2fb436936d083690068368efbc  output/bitcoin-27e63e01c-win64.zip
3f98116028552453e059e3761b9231bd1b98ba96f9322798c4f7f586779593f9  output/bitcoin-27e63e01c-x86_64-linux-gnu-debug.tar.gz
82ba38c6bb96ce0716646205a4f9054c0eeff32fe223d654f5234204afa599f4  output/bitcoin-27e63e01c-x86_64-linux-gnu.tar.gz
a8daae18536b433d05ae7972355fc99d535f2687f2e078d1a4a0893cc2c6bf5f  output/src/bitcoin-27e63e01c.tar.gz
```

Funnily enough, you'll see that the source tarballs match across Gitian+Guix

**Comment by fanquake** [comment]
Also ran a Guix build of the second to last commit here:
```bash
find output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
25aeed36ffc1eba8aff7e12f62011575e2270f07f8c25d14ffb3d0691996d919  output/bitcoin-27e63e01c-aarch64-linux-gnu-debug.tar.gz
7423aad2b0306a828b1dbe7eeb51c22c301298d9a6ae409c53546433374c8e29  output/bitcoin-27e63e01c-aarch64-linux-gnu.tar.gz
f8f027993c373698fb5f1b868769f903e246594eba69fdfe2861508828fecf0c  output/bitcoin-27e63e01c-arm-linux-gnueabihf-debug.tar.gz
6e4c079f55024348237cd55f394db1d09a1b91b8f520e64b025984d6bb7bddb0  output/bitcoin-27e63e01c-arm-linux-gnueabihf.tar.gz
7159dc6c2acabcf089b37a8c618e8558f68a44b24a733861c2b88c865c72b3f5  output/bitcoin-27e63e01c-riscv64-linux-gnu-debug.tar.gz
a85afdeff32b003f11a60831b6654a4fb1fb02b899f10392006f33fbaf2b3e4a  output/bitcoin-27e63e01c-riscv64-linux-gnu.tar.gz
273958496232e58fc4204b28947e8b10ebd0244f7016d0098b391574639aceca  output/bitcoin-27e63e01c-win-unsigned.tar.gz
49844ea00fcaebb141cc5b0e6a0a135df045dec1bb79256ef0f65e2a11d8ad1d  output/bitcoin-27e63e01c-win64-debug.zip
2c7c70a33492f17a8f35c6c7348a10ad795e85855a929196d499956b048d09a2  output/bitcoin-27e63e01c-win64-setup-unsigned.exe
8985a7d8d0b8511c2e36e050c0f006346ef66c2fb436936d083690068368efbc  output/bitcoin-27e63e01c-win64.zip
3f98116028552453e059e3761b9231bd1b98ba96f9322798c4f7f586779593f9  output/bitcoin-27e63e01c-x86_64-linux-gnu-debug.tar.gz
82ba38c6bb96ce0716646205a4f9054c0eeff32fe223d654f5234204afa599f4  output/bitcoin-27e63e01c-x86_64-linux-gnu.tar.gz
a8daae18536b433d05ae7972355fc99d535f2687f2e078d1a4a0893cc2c6bf5f  output/src/bitcoin-27e63e01c.tar.gz
```

**Comment by dongcarl** [comment]
This should be good to go, let me know if there are other things to do.

**Comment by dongcarl** [comment]
Performed builds of bfe1ba2f5b36056e0c41edf8206b93d3d83098df locally:

Gitian Linux:

```
e7509ca8cbc8b72b3796a34211243037654c70fbe29bf9b16ea1c34b36996be3  bitcoin-bfe1ba2f5b36-aarch64-linux-gnu-debug.tar.gz
54289277f8fbbc7dba36d32e81988426a2be58263aeff23f4c322f7950f42610  bitcoin-bfe1ba2f5b36-aarch64-linux-gnu.tar.gz
ad23b89b11b0eec4f9b44d04214c2e36bec01001a46093c7952e4c66a1d44dc0  bitcoin-bfe1ba2f5b36-arm-linux-gnueabihf-debug.tar.gz
c6631b3b40e073e3c65c3c3afca17e3e83dd70f124a63ed5c190775f5738f379  bitcoin-bfe1ba2f5b36-arm-linux-gnueabihf.tar.gz
a9bc8ba2f6d7b902391ab0ec1f52833dc940fda4e339885a8ec232b6798cbbaa  bitcoin-bfe1ba2f5b36-riscv64-linux-gnu-debug.tar.gz
6ccd7017ea5c5f5c2be6c0e4e244a90e6c1af710bf6124d8714f5ff51cad5964  bitcoin-bfe1ba2f5b36-riscv64-linux-gnu.tar.gz
c2efc0f69db9ab382d36addacddaa7787f9cbcfb06b4e382e345e62432e46bb6  bitcoin-bfe1ba2f5b36-x86_64-linux-gnu-debug.tar.gz
f5149e26b6b769e9a1399dab647b2fea3b27c8e9468ab6a360c2e1a69eb9fa1c  bitcoin-bfe1ba2f5b36-x86_64-linux-gnu.tar.gz
1b3994890226c3cccb41cd986b2af6ef973a7470b60ab46a9496b5664ab10db0  src/bitcoin-bfe1ba2f5b36.tar.gz
2c31afb99ab451bab3027bfc43c85a1b95a63e7e4534774e2a357d286665d4f9  bitcoin-core-linux-0.21-res.yml
```

Gitian Win:

```
b1987bbfb438f7583c66ea0e86ce9828c417cf5320d7b3f2cd6082d9a63b52f1  bitcoin-bfe1ba2f5b36-win-unsigned.tar.gz
be325defbee9d508156bc0cc6d04b634946249996123b1a555776ab0c09e4814  bitcoin-bfe1ba2f5b36-win64-debug.zip
c5b2aaa41b4b50c658724bde35cd823dcb879a2bebf095abfa5d7f43729b65a7  bitcoin-bfe1ba2f5b36-win64-setup-unsigned.exe
41be646fe5e5caf324f5d01eb79fdf4a4a58b37cebc5ada6af53214718a6bdcd  bitcoin-bfe1ba2f5b36-win64.zip
1b3994890226c3cccb41cd986b2af6ef973a7470b60ab46a9496b5664ab10db0  src/bitcoin-bfe1ba2f5b36.tar.gz
5b6a9cbfe65852375a910c4f1bb7724f79a39bf6cba7a3c7ea114ab16417304e  bitcoin-core-win-0.21-res.yml
```

All available Guix architectures:

```
$ find output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
00b0ba2bdf229d659c0326f865a9df99460b51c7b9451f8ae90f2733053a6183  output/bitcoin-bfe1ba2f5b36-aarch64-linux-gnu-debug.tar.gz
9b882e8a282fe591edce52af1ee19c270df6a7ec89593f27e79490630fe6f33c  output/bitcoin-bfe1ba2f5b36-aarch64-linux-gnu.tar.gz
564ec3c9ad00fca0e02f5abc6baf98a2794a7cc894218bd934a0ca45a4f78953  output/bitcoin-bfe1ba2f5b36-arm-linux-gnueabihf-debug.tar.gz
442413ddfcd3d5906f230a662b96f50d1d271b855be21429fb96d068842c7b1b  output/bitcoin-bfe1ba2f5b36-arm-linux-gnueabihf.tar.gz
7a8c76b3046a8c8534346c73b517dd4120aa3d47dc4ffb66455cb8d1e723a7d6  output/bitcoin-bfe1ba2f5b36-riscv64-linux-gnu-debug.tar.gz
c2893109849d9bd1fdd9fe381b9f75d6047fe320b038647de0ea5f62055c8c60  output/bitcoin-bfe1ba2f5b36-riscv64-linux-gnu.tar.gz
ce3a66d52d2a1eca761be7f0888ecf1c1571b91cc9f7b25764f70f088323c8b9  output/bitcoin-bfe1ba2f5b36-win-unsigned.tar.gz
d49cd3847560a2109c38eead3287af211e6f3e7901c6458ee78eba87db66aef2  output/bitcoin-bfe1ba2f5b36-win64-debug.zip
4d2b9be90dac86106aa4c66ad34080d317ed50cf276de8a088d046e603148025  output/bitcoin-bfe1ba2f5b36-win64-setup-unsigned.exe
34edf08ece840c9b8077ca93f8cb89fe5a7bd0cab0f7d9864aa050730b1998a3  output/bitcoin-bfe1ba2f5b36-win64.zip
c1f65264509257bd3b13159e9cf0304e4b8231ef492dc21c5fc603d1b8567aae  output/bitcoin-bfe1ba2f5b36-x86_64-linux-gnu-debug.tar.gz
a965242e1ab129ded88a88f8109e499b60f75864f2966edce9a57127fc8e2736  output/bitcoin-bfe1ba2f5b36-x86_64-linux-gnu.tar.gz
1b3994890226c3cccb41cd986b2af6ef973a7470b60ab46a9496b5664ab10db0  output/src/bitcoin-bfe1ba2f5b36.tar.gz
```

**Comment by MarcoFalke** [comment]
@dongcarl , @fanquake This conflicts with 

*     #18818 (Fix release tarball generated by gitian by luke-jr)

which is tagged 0.20.0. Just saying, to be sure we don't accidentally close doors behind us.

**Comment by dongcarl** [comment]
@MarcoFalke We can do this one of 2 ways:

1. Merge #18818 first, ship 0.20.0, then rebase this on top
2. Merge this first, then rebase #18818 on top, then ship 0.20.0

Personally, I'd prefer the second option, as this PR is quite well-tested now, and fixes the broken Linux Guix builds (which we now ship with release tarballs since the move to `git-archive`). I don't think rebasing #18818 on top of this will lead to unsolvable problems.

**Comment by fanquake** [comment]
ACK bfe1ba2f5b36056e0c41edf8206b93d3d83098df - I agree with Carl, and am going to merge this. I'd like for Linux Guix builds to be working again, and we can rebase #18818.

**Comment by MarcoFalke** [comment]
Sorry, I don't follow. Should this pull be marked for backport then?

File: contrib/guix/libexec/build.sh [hunk 2]
```diff
@@ -264,7 +259,7 @@ export PATH="${BASEPREFIX}/${HOST}/native/bin:${PATH}"
                 cp "${DISTSRC}/doc/README_windows.txt" "${DISTNAME}/readme.txt"
                 ;;
             *linux*)
-                cp "${DISTSRC}/doc/README.md" "${DISTNAME}/"
+                cp "${DISTSRC}/README.md" "${DISTNAME}/"
```
Comment (MarcoFalke) [inline contrib/guix/libexec/build.sh:262]: why is this changed?

Comment (MarcoFalke) [reply contrib/guix/libexec/build.sh:262]: Oh, I see c4a3c25ba11b20871699e570fc7b98a7b8472503

Comment (MarcoFalke) [reply contrib/guix/libexec/build.sh:262]: Neither README make sense. The top level one talks about testing and quality assurance, which is meta for the tarball. The doc README has some dead links, but at least gives some more context. :shrug:
